### PR TITLE
[Feature] Refactor test framework & test kuberay-operator chart with configuration framework

### PR DESCRIPTION
**Pull request by kevin85421**
## Why are these changes needed?
1. Test operator chart: This PR uses the kuberay-operator chart to install KubeRay operator. Hence, the operator chart is tested.

2. Refactor: `class CONST` and `class KubernetesClusterManager` should be singleton classes. However, the singleton design pattern is not encouraged, so we need to consider it thoroughly before we convert these two classes into singleton classes.

3. Refactor: Replace `os` with `subprocess`. The following paragraph is from [Python's official documentation](https://docs.python.org/3/library/os.html#os.system).

    > The [subprocess](https://docs.python.org/3/library/subprocess.html#module-subprocess) module provides more powerful facilities for spawning new processes and retrieving their results; using that module is preferable to using this function. See the [Replacing Older Functions with the subprocess Module](https://docs.python.org/3/library/subprocess.html#subprocess-replacements) section in the [subprocess](https://docs.python.org/3/library/subprocess.html#module-subprocess) documentation for some helpful recipes.

4. Skip `test_kill_head` due to
    * #638 
    * #634.

6. Refactor: Replace all existing k8s api clients with `K8S_CLUSTER_MANAGER`.

7. Refactor and relieve flakiness of `test_ray_serve_work`
    * `working_dir` is out-of-date (See this [comment](https://github.com/ray-project/kuberay/pull/698#discussion_r1017077058) for more details), but the tests pass sometimes due to the error of the original test logic. => Solution: Update `working_dir` in `ray-service.yaml.template`.
    * To elaborate, the error of the test logic mentioned above is that it only checks the exit code rather than STDOUT.
    * When Pods are READY and RUNNING, RayService still needs tens of seconds to be ready for serving requests. The `        time.sleep(60)` function is a workaround, and should be removed when https://github.com/ray-project/kuberay/pull/730 is merged.
    * Remove NodePort service in `RayServiceTestCase`. Use a `curl` Pod to communicate with Ray via ClusterIP service directly. Originally, using Docker container with `network_mode='host'` and NodePort service is very weird for me.

8. Refactor: remove useless RayService template `ray-service-cluster-update.yaml.template` and `ray-service-serve-update.yaml.template`. The original buggy test logic only checks the exit code rather than the `STDOUT` of the curl commands. Hence, the different templates are useless in `RayServiceTestCase`.

9. Refactor: Because APIServer is not tested by any test case, remove everything related to APIServer docker image in the compatibility test.

Follow up issues (@jasoonn and I will finish them):
* #767 
* #766
* #764
* #762 
* #763 
* #771

## Related issue number
Part of #184 

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [x] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

## Flakiness (no retry)
* [445b61f](https://github.com/ray-project/kuberay/pull/759/commits/445b61f900310a2f50a20cf1d35c0c2ec9e656eb): Pass
* [d599c71](https://github.com/ray-project/kuberay/pull/759/commits/d599c714cc6f3ca2618412e5854d73601e0a11e3): Pass
* [d977b2b](https://github.com/ray-project/kuberay/pull/759/commits/d977b2b1295a5e471223461b22a16e9bddcb28fe): Fail
  * Ray client cannot connect to Ray cluster (`test_detach_actor_1.py`) => This is weird, and may be caused by the weird pattern of Docker container and Kubernetes NodePort service.
  * `curl: (7) Failed to connect to rayservice-sample-serve-svc.default.svc.cluster.local port 8000: Connection refused
command terminated with exit code 7` => This should be fixed by #730  
* [829195b](https://github.com/ray-project/kuberay/pull/759/commits/829195b2bd5fff6b29e849b7fae32bd3146cc89a): Pass
* [78849d6](https://github.com/ray-project/kuberay/pull/759/commits/78849d69ed70f4c9456ec4b6340ccb07d9296ed4): Fail
  * `Exception: There was an exception during the execution of test_ray_serve_1.py.` => May be caused by Docker container and NodePort service.
* [66107f9](https://github.com/ray-project/kuberay/pull/759/commits/66107f9b4c783fd1bc094ff2789bd0ada17a94a2): Fail 
  * `Exception: There was an exception during the execution of test_ray_serve_1.py.`
  *  `curl: (7) Failed to connect to rayservice-sample-serve-svc.default.svc.cluster.local port 8000: Connection refused
command terminated with exit code 7` => This should be fixed by #730

### CODE REVIEW

**Comment by kevin85421** [comment]
cc @jasoonn Would you mind reviewing this PR? Thank you!

**Review by jasoonn — COMMENTED** [review]
@kevin85421 Thanks for the contribution!

File: tests/compatibility-test.py [hunk 1]
```diff
@@ -36,12 +42,15 @@ def setUpClass(cls):
...
-        utils.apply_kuberay_resources(images, kuberay_operator_image, kuberay_apiserver_image)
-        utils.create_kuberay_cluster(BasicRayTestCase.cluster_template_file,
+        K8S_CLUSTER_MANAGER.delete_kind_cluster()
+        K8S_CLUSTER_MANAGER.create_kind_cluster()
+        image_dict = {
+            CONST.RAY_IMAGE_KEY: ray_image,
+            CONST.OPERATOR_IMAGE_KEY: kuberay_operator_image
+        }
+        operator_manager = OperatorManager(image_dict)
+        operator_manager.prepare_operator()
```
Comment (jasoonn) [inline tests/compatibility-test.py:52]: It looks like line 47 to 52 appears multiple times(BasicRayTestCase, RayFTTestCase, and RayServiceTestCase). Maybe we can combine them as a function?

File: tests/compatibility-test.py [hunk 2]
```diff
@@ -138,38 +147,42 @@ class RayFTTestCase(unittest.TestCase):
...
         # shell_assert_success('kubectl wait --for=condition=Ready pod/$(kubectl get pods -A | grep -e "-head" | awk "{print \$2}") --timeout=900s')
         # make sure both head and worker pods are ready
-        rtn = utils.shell_run(
-            'kubectl wait --for=condition=ready pod -l rayCluster=raycluster-compatibility-test --all --timeout=900s')
+        rtn = shell_subprocess_run(
+                'kubectl wait --for=condition=ready pod -l rayCluster=raycluster-compatibility-test --all --timeout=900s')
         if rtn != 0:
-            utils.shell_run('kubectl get pods -A')
-            utils.shell_run(
+            shell_subprocess_run('kubectl get pods -A')
```
Comment (jasoonn) [inline tests/compatibility-test.py:180]: Because the default check argument is true in ```shell_subprocess_run```, it will raise an error when the return code of execution is not a 0. Hence, this part of program will not be executed at anytime.

File: tests/compatibility-test.py [hunk 3]
```diff
@@ -264,75 +269,49 @@ def test_detached_actor(self):
...
-            ray_version, ray_image)
-        time.sleep(5)
-        curl_cmd = 'curl  -X POST -H \'Content-Type: application/json\' localhost:8000 -d \'["MANGO", 2]\''
-        utils.wait_for_condition(
-            lambda: utils.shell_run(curl_cmd) == 0,
-            timeout=180,
-        )
-
+        """Create a RayService, send a request to RayService via `curl`, and compare the result."""
+        cr_event = utils.create_ray_service(
```
Comment (jasoonn) [inline tests/compatibility-test.py:293]: Do you think we need to add a test for updating RayService?

File: tests/compatibility-test.py (hunk 1 above)
Comment (jasoonn) [inline tests/compatibility-test.py:52]: It looks like line 45 to 52 appears multiple times(BasicRayTestCase, RayFTTestCase, and RayServiceTestCase). Maybe we can combine them as a function?

File: .github/workflows/actions/configuration/action.yaml [hunk 4]
```diff
@@ -48,29 +48,13 @@ runs:
...
-      docker images ls -a
-    shell: bash
-
   - name: Run configuration tests for sample YAML files.
     # compatibility test depends on operator & apiserver images built in previous steps.
     run: |
-      pushd manifests/base/
-      kustomize edit set image kuberay/operator=kuberay/operator:${{ steps.vars.outputs.sha_short }}
-      kustomize edit set image kuberay/apiserver=kuberay/apiserver:${{ steps.vars.outputs.sha_short }}
-      popd
```
Comment (jasoonn) [inline .github/workflows/actions/configuration/action.yaml:69]: Except here and [compatibility/action.yaml](https://github.com/ray-project/kuberay/pull/759/files/829195b2bd5fff6b29e849b7fae32bd3146cc89a#diff-88197d1eee4d640617cb62d8517a0490a2406e4248d9b14214047d299f61a7fe), does kustomize be utilized in other places? If not, I think we can remove kustomize dependency in [ray-operator/DEVELOPMENT.md](https://github.com/ray-project/kuberay/blob/master/ray-operator/DEVELOPMENT.md).

File: tests/compatibility-test.py [hunk 5]
```diff
@@ -138,38 +147,42 @@ class RayFTTestCase(unittest.TestCase):
...
-        rtn = utils.shell_run(
-            'kubectl wait --for=condition=ready pod -l rayCluster=raycluster-compatibility-test --all --timeout=900s')
+        rtn = shell_subprocess_run(
+                'kubectl wait --for=condition=ready pod -l rayCluster=raycluster-compatibility-test --all --timeout=900s')
         if rtn != 0:
-            utils.shell_run('kubectl get pods -A')
-            utils.shell_run(
+            shell_subprocess_run('kubectl get pods -A')
+            shell_subprocess_run(
                 'kubectl describe pod $(kubectl get pods | grep -e "-head" | awk "{print \$1}")')
```
Comment (jasoonn) [inline tests/compatibility-test.py:182]: Because the default check argument is true in shell_subprocess_run, it will raise an error when the return code of execution is not a 0. Hence, the program will not enter the section below```if rtn != 0``` .

File: tests/compatibility-test.py [hunk 6]
```diff
@@ -264,75 +269,49 @@ def test_detached_actor(self):
...
-
+        """Create a RayService, send a request to RayService via `curl`, and compare the result."""
+        cr_event = utils.create_ray_service(
+            RayServiceTestCase.service_template_file, ray_version, ray_image)
+        # When Pods are READY and RUNNING, RayService still needs tens of seconds to be ready
+        # for serving requests. This `sleep` function is a workaround, and should be removed
+        # when https://github.com/ray-project/kuberay/pull/730 is merged.
+        time.sleep(60)
+        cr_event.rulesets = [RuleSet([CurlServiceRule()])]
+        cr_event.check_rule_sets()
```
Comment (jasoonn) [inline tests/compatibility-test.py:300]: Do you think we need to add a test for updating RayService as well?

File: tests/framework/prototype.py [hunk 7]
```diff
@@ -38,41 +61,82 @@ def search_path(yaml_object, steps, default_value = None):
...
+                # Only pull the image from DockerHub when the image does not
+                # exist in the local docker registry.
+                image = self.docker_image_dict[key]
+                if shell_subprocess_run(
+                        f'docker image inspect {image} > /dev/null', check = False) != 0:
+                    shell_subprocess_run(f'docker pull {image}')
+                else:
+                    logger.info("Image %s exists", image)
+
+        download_images()
```
Comment (jasoonn) [inline tests/framework/prototype.py:126]: I am wondering why you define the inner function download_images() here instead of writing them down directly.

File: tests/compatibility-test.py (hunk 1 above)
Comment (kevin85421) [reply tests/compatibility-test.py:52]: We decided not to combine them as a function after an offline discussion. Reasons: 

(1) Because defining a new function will couple both Kubernetes cluster and operator abstractions, it's hard to think of an easy-to-understand abstraction for the function consisting of L45 to L52. It will hurt the readability at this moment. 

(2) We may need to execute something between KinD cluster creation and operator creation. Take #750 as an example, create a new namespace with `restricted` Pod security standard.

File: tests/framework/prototype.py (hunk 7 above)
Comment (kevin85421) [reply tests/framework/prototype.py:126]: We always load Docker images into a KinD cluster after `download_images()`. Hence, defining `download_images()` as an inner function can reduce the cognitive load on developers who use the module. To elaborate, they did not need to know that `kind load ...` must be called after `download_images()`.

Comment (kevin85421) [reply tests/framework/prototype.py:126]: It's a good chance to read Chapter 5 of the book "A Philosophy of Software Design".

File: .github/workflows/actions/configuration/action.yaml (hunk 4 above)
Comment (kevin85421) [reply .github/workflows/actions/configuration/action.yaml:69]: Good catch! I will update it.

File: tests/compatibility-test.py (hunk 5 above)
Comment (kevin85421) [reply tests/compatibility-test.py:182]: Good catch! I will update it. By the way, this test is skipped, and is highly possible to rewrite the test logic.

File: .github/workflows/actions/configuration/action.yaml (hunk 4 above)
Comment (kevin85421) [reply .github/workflows/actions/configuration/action.yaml:69]: Update https://github.com/ray-project/kuberay/pull/759/commits/78849d69ed70f4c9456ec4b6340ccb07d9296ed4

File: tests/compatibility-test.py (hunk 5 above)
Comment (kevin85421) [reply tests/compatibility-test.py]: Updated.

File: tests/compatibility-test.py (hunk 6 above)
Comment (kevin85421) [reply tests/compatibility-test.py:300]: Maybe we should have a test related to updating RayService in the future. Currently, I want to make sure the simplest test is stable enough. The original test logic is too weird. I will discuss with @shrekris-anyscale and @architkulkarni to design a thorough test plan for both RayService and RayJob.

File: tests/compatibility-test.py [hunk 8]
```diff
@@ -8,7 +8,14 @@
 import docker
 
 from kuberay_utils import utils
-from kubernetes import client, config
+from framework.prototype import (
+    CONST,
```
Comment (DmitriGekhtman) [inline tests/compatibility-test.py:12]: All of the objects that are not directly related to the prototype test framework should probably live in the utils file.

The objects that should be moved include `CONST`, `K8S_CLUSTER_MANAGER`, `OperatorManager`, and `shell_subprocess_run`.

File: tests/framework/prototype.py [hunk 9]
```diff
@@ -4,18 +4,41 @@
...
+    # Docker images
+    OPERATOR_IMAGE_KEY = "kuberay-operator-image"
+    RAY_IMAGE_KEY = "ray-image"
+
+    # Kubernetes API clients
+    K8S_CR_CLIENT_KEY = "k8s-cr-api-client"
+    K8S_V1_CLIENT_KEY = "k8s-v1-api-client"
+
+    # Paths
+    REPO_ROOT = "/" + "/".join(os.path.abspath(os.path.dirname(__file__)).split('/')[1:-2]) + "/"
```
Comment (DmitriGekhtman) [inline tests/framework/prototype.py:31]: Could you simplify this expression?
You might want to use the pathlib module for this.

File: tests/framework/prototype.py [hunk 10]
```diff
@@ -4,18 +4,41 @@
...
+    HELM_CHART_ROOT = REPO_ROOT + "helm-chart"
```
Comment (DmitriGekhtman) [inline tests/framework/prototype.py:32]: Use os.path.join or pathlib to form paths.

File: tests/framework/prototype.py [hunk 11]
```diff
@@ -38,41 +61,82 @@ def search_path(yaml_object, steps, default_value = None):
...
+        config.load_kube_config()
+        self.k8s_client_dict.update({
+            CONST.K8S_V1_CLIENT_KEY: client.CoreV1Api(),
+            CONST.K8S_CR_CLIENT_KEY: client.CustomObjectsApi()
+        })
+
+    def check_cluster_exist(self) -> bool:
+        """Check whether KinD cluster exists or not"""
+        return shell_subprocess_run("kubectl cluster-info --context kind-kind", check = False) == 0
+K8S_CLUSTER_MANAGER = KubernetesClusterManager()
```
Comment (DmitriGekhtman) [inline tests/framework/prototype.py:93]: There should be a space (or two) after the class. 
Would you mind running a linter?

File: tests/framework/prototype.py [hunk 12]
```diff
@@ -4,18 +4,41 @@
...
 
-# Utility functions
+# Global variables
 logger = logging.getLogger(__name__)
 logging.basicConfig(
     format='%(asctime)s,%(msecs)d %(levelname)-8s [%(filename)s:%(lineno)d] %(message)s',
     datefmt='%Y-%m-%d:%H:%M:%S',
     level=logging.INFO)
 
+class CONST(object):
```
[the remark this answers is not in the document]
Comment (DmitriGekhtman) [reply tests/framework/prototype.py:19]: Maybe call this class `TEST_CONSTANTS`?

**Comment by DmitriGekhtman** [comment]
> Refactor: remove useless RayService template ray-service-cluster-update.yaml.template and ray-service-serve-update.yaml.template. The original buggy test logic only checks the exit code rather than the STDOUT of the curl commands. Hence, the different templates are useless in RayServiceTestCase.

Could you make a note in the code and on the RayService test issue that
- The previous logic for testing updates was problematic
- We need to test updates

**Comment by DmitriGekhtman** [comment]
Some minor style complaints above.

Overall this looks great! The KubeRay CI should be much greener after these changes.

File: tests/framework/prototype.py (hunk 11 above)
Comment (kevin85421) [reply tests/framework/prototype.py:93]: My pylint is always running. Would you mind explaining "There should be a space (or two) after the class." more? Do you mean add newlines between L92 and L93? Thank you!

File: tests/framework/prototype.py (hunk 9 above)
Comment (kevin85421) [reply tests/framework/prototype.py]: Updated https://github.com/ray-project/kuberay/pull/759/commits/b2a23897408b00122fe5729b8190a850c63ff82b.

File: tests/framework/prototype.py (hunk 10 above)
Comment (kevin85421) [reply tests/framework/prototype.py]: Updated https://github.com/ray-project/kuberay/pull/759/commits/b2a23897408b00122fe5729b8190a850c63ff82b.

File: tests/compatibility-test.py (hunk 8 above)
Comment (kevin85421) [reply tests/compatibility-test.py:12]: Is this necessary in this PR? I am not familiar with `import` in Python. If we need to move them into a separate module, I need time to understand the import mechanism of Python. If it is not required in this PR, I can open an issue to track this issue.

File: tests/framework/prototype.py (hunk 11 above)
Comment (DmitriGekhtman) [reply tests/framework/prototype.py:94]: Yeah, that's what I meant. Maybe our linters are different.

Instantiating the class in the line after its defined makes enough sense to me!  Don't mind me.

File: tests/compatibility-test.py (hunk 8 above)
Comment (DmitriGekhtman) [reply tests/compatibility-test.py:12]: It's not so important for this PR and can be done later!
Utilities used by all tests (like `shell_subprocess_run`) should probably be collected in a "utilities" file.

File: tests/framework/prototype.py [hunk 13]
```diff
@@ -28,9 +28,9 @@ class CONST(object):
     K8S_V1_CLIENT_KEY = "k8s-v1-api-client"
 
     # Paths
-    REPO_ROOT = "/" + "/".join(os.path.abspath(os.path.dirname(__file__)).split('/')[1:-2]) + "/"
-    HELM_CHART_ROOT = REPO_ROOT + "helm-chart"
-    DEFAULT_KIND_CONFIG = REPO_ROOT + "tests/framework/config/kind-config.yaml"
+    REPO_ROOT = Path(__file__).absolute().parent.parent.parent
```
Comment (DmitriGekhtman) [inline tests/framework/prototype.py:31]: I think you can do .parents[3].
But .parent.parent.parent is fine too.

**Comment by kevin85421** [comment]
> > Skip test_kill_head.
> 
> Could you give context on that decision? Is the test complete nonsense or do we need to follow up -- do we need to fix it or write a test that covers similar logic?

* We need to rewrite the test (for example, remove `kubectl wait`).
* I skipped this test due to 
  * #638 
  * #634

File: tests/framework/prototype.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (kevin85421) [reply tests/framework/prototype.py]: > You don't need to inherit from object in Python 3.

Good catch! Updated.

> Maybe call this class TEST_CONSTANTS?

The name is a bit long.

File: tests/compatibility-test.py (hunk 8 above)
Comment (kevin85421) [reply tests/compatibility-test.py:12]: Open an issue #774 to track the progress.

**Review by DmitriGekhtman — APPROVED** [review]
!!!

**Comment by kevin85421** [comment]
> * I skipped this test due to
>   
>   * [[Bug] Head pod is deleted rather than restarted when gcs_server on head pod is killed. #638](https://github.com/ray-project/kuberay/issues/638)
>   * [[Bug] Worker pods crash unexpectedly when gcs_server on head pod is killed  #634](https://github.com/ray-project/kuberay/issues/634)

@DmitriGekhtman These are wrong. These two issues are related to `test_detached_actor` and `test_ray_serve`. However, I observed that sometimes workers crash when the head pod is killed.
